py-polars/polars/datatypes.py

```diff
@@ -218,6 +221,8 @@ def __hash__(self) -> int:
 class Duration(DataType):
     """Time duration/delta type."""
 
+    tu: TimeUnit | None = None
```

Comment:
Is there a reason we allow `None` here? I would think a `Duration` _needs_ a `TimeUnit`

Reply:
Ahh, true - conflated the Datetime/Duration "tu" units.

Reply:
Ahh, I double-checked; it actually doesn't, as it implicitly uses [μs] if not set.
```python
df = pl.DataFrame({"a": [100_000_000, 200_000_000]})
df.select( pl.col("a").cast(pl.Duration) )

# shape: (2, 1)
# ┌──────────────┐
# │ a            │
# │ ---          │
# │ duration[μs] │
# ╞══════════════╡
# │ 1m 40s       │
# ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
# │ 3m 20s       │
# └──────────────┘
```
We could make it an _explicit_ default, but it's currently not necessary (setting it at the class level just guarantees that there is a `tu` attribute, even if the caller didn't set it via init).

Reply:
Ok. I do think we should avoid setting different defaults in different places, though. For example,

```python
class Datetime(DataType):
    """Calendar date and time type."""

    tu: TimeUnit | None = None
    tz: str | None = None

    def __init__(self, time_unit: TimeUnit = "us", time_zone: str | None = None):
        ...
```

so the default is `"us": TimeUnit` in the `__init__` but on the class it's defined differently. I would recommend 

```python
class Datetime(DataType):
    """Calendar date and time type."""

    tu: TimeUnit
    tz: str | None

    def __init__(self, time_unit: TimeUnit = "us", time_zone: str | None = None):
        ...
```

We then can keep `__init__` as the sole provider for defaults. It is also arguably more "correct", since we should never have `tu: None` as an attribute.

Same argument for `Duration`

Reply:
Not a fan of public class variables spontaneously coming into existence only on `__init__`, as the calling code inevitably becomes awkward: `getattr(series,'tu',None)` vs `series.tu`. The "real" underlying issue is that we currently use class instances as if they were specialisations of the class; I'm tempted to revisit that later... ;)

Reply:
> The "real" underlying issue is that we currently use class instances as if they were specialisations of the class; I'm tempted to revisit that later... ;)

Completely agree, it really confuses mypy (not to mention, my brain).

> Not a fan of public class variables spontaneously coming into existence only on `__init__`, as the calling code inevitably becomes awkward: getattr(series.dtype,'tu',None) vs series.dtype.tu.

I'm sorry to keep pushing back on this but `tu` and `tz` aren't class variables -- if they are, then they should be annotated with `ClassVar`.  They are instance variables, annotated in the body of the class for readability (see [PEP-526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations)).

Reply:
Perhaps it would help me if you could elaborate on how the calling code becomes awkward with my approach

Reply:
Thank you for that! Ok, I agree adding the `None` annotation makes sense. 

This does seem to be a symptom of our strange handling of instance vs. class dtypes. I too would like that to see cleaned up at some point

Reply:
I solved the issue quite nicely a few years ago with an earlier iteration of the data API I built for JPMorgan (which has to handle all manner of types, subtypes, units of types, etc, for multiple flavours of backend - relational, streaming, frames, and so on); quite likely I can retrofit some of the same concepts here in simpler form. The real trick will be to do so without too much disruption... :)